### ISSUE TITLE
Retrying after database error causes Hibernate exception

### ISSUE THREAD

**Issue by hprange** [quarkusio/quarkus#16198]
## Describe the bug
My project has a JPA entity annotated with `@UniqueConstraint` to guarantee the application won't persist duplicate data. One of its endpoints checks the presence of a key in the database and, if missing, inserts a new record. Under highly concurrent circumstances, the application may trigger unique constraints while saving new data. I added the `@Retry` annotation to the resource method to make REST API invocations more fault-tolerant. However, Hibernate throws an exception when Quarkus attempts to retry the request.

### Expected behavior
Quarkus should attempt to execute the same method again and return the previously saved data.

### Actual behavior
Quarkus attempts to execute the same method but throws an `org.hibernate.AssertionFailure`.

```
ERROR [org.hib.AssertionFailure] (executor-thread-1) HHH000099: an assertion failure occurred (this may indicate a bug in Hibernate, but is more likely due to unsafe use of the session): org.hibernate.AssertionFailure: null id in org.acme.Car entry (don't flush the Session after an exception occurs)
```

## To Reproduce

Clone [this](https://github.com/hprange/retry-hibernate-error) repository and run the tests.

### Output of `uname -a` or `ver`
Darwin Kernel Version 20.3.0:root:xnu-7195.81.3~1/RELEASE_X86_64 x86_64

### Output of `java -version`
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.10+9, mixed mode)

### Quarkus version or git rev
1.13.0.FINAL

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
Apache Maven 3.6.3

**Comment by gavinking** [comment]
When an exception propagates out of the Hibernate session, the session is considered in an inconsistent state and must be discarded. You can't just retry things using the same session.

**Comment by hprange** [comment]
Thanks for replying so fast, @gavinking. Shouldn't Quarkus discard the Hibernate session automatically in this case? If not, how can I tell Quarkus to discard the Hibernate session before making a new attempt to execute the resource method?

**Comment by Sanne** [comment]
@hprange yes I think you're right, this needs to be handled by the code implementing the `@Retry` annotation bindings.

I'm not familiar with that though. @radcortez could you suggest someone to assign this to?

**Comment by radcortez** [comment]
> @hprange yes I think you're right, this needs to be handled by the code implementing the `@Retry` annotation bindings.
> 
> I'm not familiar with that though. @radcortez could you suggest someone to assign this to?

@Ladicek what do you think?

**Comment by Ladicek** [comment]
SmallRye Fault Tolerance can't possibly know all the state management intricacies of all the frameworks present in the application. It also has no idea how the guarded code looks like. The basic assumption of `@Retry` is that it is safe to execute the guarded method multiple times.

That said, this could perhaps be a use case for "before retry actions" (https://github.com/smallrye/smallrye-fault-tolerance/issues/259).

**Comment by Sanne** [comment]
right, would be nice to have "before retry" actions automatically registered. What about transactions - how are they supposed to integrate with `@Retry` , is this specified somewhere?

I would expect the current transaction to be aborted and a new one started, but that's assuming the Retry annotation is on the outer endpoint. Would be a mess to auto-abort transactions from a nested method.

**Comment by Ladicek** [comment]
There is no specification on how `@Retry` should interact with JPA or transactions. MicroProfile Fault Tolerance is mainly designed for guarding straightforward request/response network interactions.

I also realized another thing. If I understand correctly, Hibernate sessions are managed by Quarkus and are assumed to span the whole request. If we were to discard them in the middle of request processing and replace with fresh ones, you would easily end up with detached entities (and perhaps more troubles? I don't really understand Hibernate).

I _guess_ this could all be solved in a straightforward way if, as @Sanne points out, we were to retry the entire request processing pipeline. That is currently probably equivalent to putting `@Retry` on the outermost resource method, but I recall there were ideas about injecting entities into resource methods to simplify some common tasks, Panache-style. In such case, putting `@Retry` on such method, even if it were the outermost resource method, would _not_ be equivalent to retrying the entire request processing pipeline. Which leads me to think this should perhaps be an entirely separate feature?

**Comment by Sanne** [comment]
Thanks for clarifying @Ladicek .

I must admit I'm not a fan of these "fault-tolerant" annotations, if it's a dumb "try again in a loop" thing that seems not very helpful and not motivating people to fix the underlying issues.

We probably need to keep this simple for now - sorry @hprange , seems like you shouldn't use it in this case.

**Comment by Ladicek** [comment]
I agree that blindly retrying in case of an error is, itself, an error. `@Retry` is to be used very sparingly, though for some reason, a lot of people reach for it very soon :-) It's something I plan to add to the documentation when I expand it (https://github.com/smallrye/smallrye-fault-tolerance/issues/397).

**Comment by Sanne** [comment]
Coming back to this a bit late as I'm reviewing old issues :)

I think we agreed that `@Retry` needs to be used with care on bean methods; for example it would be a design error to try reusing an EntityManager (or a Transaction) but "retry" a method which affects its state partially.

But this issue description specifically mentions:

> I added the `@Retry` annotation to the resource method to make REST API invocations more fault-tolerant.

In this particular case I'd say that @hprange 's expectations seem reasonable in expecting the whole request to be re-processed - essentially we should re-start the request context and start over.

This should work for simple API calls - not sure if it's doable for streams or larger payloads which might have been consumed during processing? Would be nice if we were able to detect when a `@Retry` annotation is being placed on an endpoint that doesn't support it.

**Comment by Ladicek** [comment]
Thinking about it again, I guess we could only possibly guarantee "reprocessability" for what the HTTP specification calls "safe methods" (`GET` and `HEAD`, mostly). Anything else is hit or miss due to reasons you mention (do we wanna buffer the HTTP entity body, for example?), including the original issue here (which I guess is a `PUT` or `POST`).
